 

### 실습 1. Excel 파일 불러오기

 

해야 할 일

 

pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

 

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.

 

In [7]:
import pandas as pd  # pandas 라이브러리를 불러오고, 이후 pd라는 이름으로 사용합니다

df = pd.read_excel('교보문고_종합_베스트셀러_상품리스트.xlsx')  # 엑셀 파일을 읽어와 df라는 이름의 표(DataFrame)로 저장합니다

df.head()  # df의 맨 위 5개 행을 보여줍니다 (기본값이 5행입니다)

c:\dev\llm-data-analysis-course\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


### 결과 요약

- 총 5개의 열(순위, 상품코드, 판매상품ID, 상품명, 정가, 판매가, 할인율, 적립율, 적립예정포인트, 인물, 출판사, 발행(출시)일자, 분야)로 구성된 데이터를 불러왔습니다.
- `df.head()`로 확인한 상위 5개 행은 모두 '소설' 또는 '정치/사회' 분야의 베스트셀러 도서입니다.
- 예: 1위 『소년이 온다』(한강, 창비), 3위 『결국 국민이 합니다』(이재명, 오마이북) 등.
- 참고: 실행 시 `openpyxl` 관련 `UserWarning`(기본 스타일 적용 안내)이 출력되었으나, 데이터 로딩에는 문제가 없습니다.

### 실습 2. 데이터의 크기와 컬럼 확인하기

 

Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.

 

먼저 어떤 데이터인지 확인합니다.

In [8]:
# 1. 전체 행(row)과 컬럼(column) 개수 확인 (결과: (행 개수, 열 개수))
print("데이터 크기(행, 열):", df.shape)

# 2. 컬럼 이름 목록 확인
print("\n컬럼 이름:", df.columns.tolist())

# 3. 각 컬럼의 자료형 확인 (숫자형/문자형 등)
print("\n각 컬럼의 자료형:")
print(df.dtypes)

# 4. 각 컬럼의 결측치(비어있는 값) 개수 확인
print("\n컬럼별 결측치 개수:")
print(df.isna().sum())

# 5. 앞의 5행 미리보기
df.head()

데이터 크기(행, 열): (199, 13)

컬럼 이름: ['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트', '인물', '출판사', '발행(출시)일자', '분야']

각 컬럼의 자료형:
순위          int64
상품코드        int64
판매상품 ID       str
상품명           str
정가            str
판매가           str
할인율           str
적립율           str
적립예정포인트       str
인물            str
출판사           str
발행(출시)일자    int64
분야            str
dtype: object

컬럼별 결측치 개수:
순위          0
상품코드        0
판매상품 ID     0
상품명         0
정가          0
판매가         0
할인율         0
적립율         0
적립예정포인트     0
인물          0
출판사         0
발행(출시)일자    0
분야          0
dtype: int64


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


In [12]:
df.isna().sum()

순위          0
상품코드        0
판매상품 ID     0
상품명         0
정가          0
판매가         0
할인율         0
적립율         0
적립예정포인트     0
인물          0
출판사         0
발행(출시)일자    0
분야          0
dtype: int64

### 실행 결과 요약

* **데이터 규모:** 199행 × 13열[cite: 1]
* **컬럼 목록:** 순위, 상품코드, 판매상품 ID, 상품명, 정가, 판매가, 할인율, 적립율, 적립예정포인트, 인물, 출판사, 발행(출시)일자, 분야[cite: 1]
* **자료형 특징:**
  * **숫자형(int64):** `순위`, `상품코드`, `발행(출시)일자` (3개)[cite: 1]
  * **문자형(str/object):** `정가`, `판매가`, `할인율`, `적립율`, `적립예정포인트` 포함 10개[cite: 1]
  * 💡 *'정가', '판매가' 등은 쉼표(`,`)나 `%`가 섞여 문자형으로 인식되었으므로 수치형 변환(전처리) 필요*
* **결측치:** 전체 컬럼 0건 (누락 데이터 없음)[cite: 1]
* **상위 5행 데이터:** 1위 『소년이 온다』부터 5위 『급류』까지 정상 확인[cite: 1]

### 실습 3. 분석에 필요한 컬럼만 선택하기

 

원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

 

이번 과정에서는 다음 컬럼을 먼저 선택합니다.

 

순위

판매상품 ID

상품명

판매가

인물

출판사

발행(출시)일자

분야

 

왜 컬럼을 선택할까요?

 

사용하지 않는 데이터를 계속 가지고 있으면 분석 과정이 복잡해질 수 있습니다.

 

이번 과정에서 필요한 정보만 남겨 데이터의 목적을 명확하게 만듭니다.

In [13]:
# 1. 원하는 컬럼만 선택하여 새로운 데이터프레임(df_books) 생성
selected_columns = ['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야']
df_books = df[selected_columns]

# 2. 컬럼이 잘 추출되었는지 확인 (앞의 5행 출력)
df_books.head()

,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000000610612,소년이 온다,"13,500",한강,창비,20140519,소설
1,2,S000001632467,모순,"11,700",양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,"19,800",이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,"16,200",성해나,창비,20250328,소설
4,5,S000200550189,급류,"12,600",정대건,민음사,20221222,소설


### 컬럼 선택 실행 결과 요약

* **선택된 컬럼 (8개):** `순위`, `판매상품 ID`, `상품명`, `판매가`, `인물`, `출판사`, `발행(출시)일자`, `분야`
* **제외된 컬럼 (5개):** `상품코드`, `정가`, `할인율`, `적립율`, `적립예정포인트`
* **결과 확인:** `df_books` 데이터프레임이 지정한 8개 컬럼으로 정상 생성되었으며, 상위 5개 행(1위~5위 도서 정보)이 잘 출력되었습니다.

 

### 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기

 

원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.

 

이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.

 

판매상품 ID       → 판매상품ID

인물              → 저자

발행(출시)일자    → 발행일

 

나머지 컬럼은 그대로 사용해도 됩니다.

 

최종적으로 다음 컬럼을 사용할 수 있습니다.

 

순위

판매상품ID

상품명

판매가

저자

출판사

발행일

분야

In [14]:
# 1. rename() 함수를 사용하여 컬럼 이름 변경
df_books = df_books.rename(columns={
    '판매상품 ID': '판매상품ID',
    '인물': '저자',
    '발행(출시)일자': '발행일'
})

# 2. 변경된 컬럼 이름 목록 확인
df_books.columns.tolist()

['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야']

### 컬럼 이름 변경 결과 요약

* **변경 적용 내용:**
  * `판매상품 ID` → `판매상품ID` (공백 제거)
  * `인물` → `저자`
  * `발행(출시)일자` → `발행일`
* **최종 컬럼 목록 (8개):** `순위`, `판매상품ID`, `상품명`, `판매가`, `저자`, `출판사`, `발행일`, `분야`
* **결과 확인:** `rename()` 함수 실행 후 `df_books.columns.tolist()`를 통해 3개 컬럼의 이름이 의도대로 정확히 변경된 것을 확인했습니다.

### 실습 5. 결측치 확인하기

결측치는 값이 비어 있는 데이터를 의미합니다.

데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.

먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.

In [15]:
# 1. 모든 컬럼의 결측치(비어있는 값) 개수 확인
print("전체 컬럼 결측치 개수:")
print(df_books.isna().sum())

# 2. 결측치가 1개 이상 있는 컬럼만 필터링해서 보기
null_counts = df_books.isna().sum()
print("\n결측치가 있는 컬럼만 보기:")
print(null_counts[null_counts > 0])

전체 컬럼 결측치 개수:
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

결측치가 있는 컬럼만 보기:
Series([], dtype: int64)


### 결측치 확인 실행 결과 요약

* **전체 컬럼 결측치 상태:** 8개 컬럼(`순위`, `판매상품ID`, `상품명`, `판매가`, `저자`, `출판사`, `발행일`, `분야`) 모두 결측치가 0개로 확인됨
* **결측치 필터링 결과:** `null_counts > 0` 조건을 만족하는 컬럼이 없어 빈 시리즈(`Series([], dtype: int64)`)가 출력됨
* **결론:** `df_books` 데이터프레임에는 비어있는 값(NaN/Null)이 전혀 존재하지 않으므로, 결측치 대체나 삭제 작업 없이 바로 다음 단계 분석 및 전처리를 진행할 수 있음

### 실습 6. 판매가를 숫자로 바꾸기

Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다. 

16,200

17,820

7,200
 
사람이 보기에는 숫자처럼 보이지만 pandas에서는 문자열로 읽힐 수 있습니다.

문자열 상태에서는 평균, 합계 등의 숫자 계산을 제대로 할 수 없습니다.

따라서 쉼표를 제거하고 숫자 자료형으로 변환합니다.

먼저 자료형 확인

df_books["판매가"].dtype

In [16]:
# 1. '판매가' 컬럼에서 쉼표(,) 제거하기
df_books['판매가'] = df_books['판매가'].str.replace(',', '')

# 2. 문자열(str) 형태를 정수형(int)으로 변환하기
df_books['판매가'] = df_books['판매가'].astype(int)

# 3. 변환 후 '판매가' 컬럼의 자료형(dtype) 확인하기
print("판매가 컬럼 자료형:", df_books['판매가'].dtype)

# 4. 상위 5행 출력하여 값 확인하기
df_books.head()

판매가 컬럼 자료형: int64


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,20140519,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,20250328,소설
4,5,S000200550189,급류,12600,정대건,민음사,20221222,소설


### 판매가 전처리 실행 결과 요약

* **전처리 내용:** `.str.replace(',', '')`를 이용해 쉼표를 제거한 후, `.astype(int)`로 정수형 변환 완료
* **자료형 확인:** `판매가` 컬럼의 자료형이 문자열(`object`)에서 정수형(`int64`)으로 성공적으로 변경됨
* **데이터 확인:** 상위 5행 출력 결과, `'13,500'` $\rightarrow$ `13500`과 같이 쉼표가 제거된 숫자 데이터로 정상 변환된 것 확인

### 실습 7. 발행일을 날짜형으로 바꾸기

 

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

 

20260701

20260831

20020120

 

이 값은 연도-월-일 정보를 가지고 있지만 바로 날짜 계산에 사용하기 어려울 수 있습니다.

 

날짜 자료형으로 변환하면 연도, 월, 날짜 등을 쉽게 분석할 수 있습니다.


In [17]:
# 1. '발행일' 컬럼을 문자열(str)로 변환 후, YYYYMMDD 포맷에 맞게 datetime 날짜형으로 변환
df_books['발행일'] = pd.to_datetime(df_books['발행일'].astype(str), format='%Y%m%d')

# 2. 변환 후 자료형(dtype) 확인
print("발행일 컬럼 자료형:", df_books['발행일'].dtype)

# 3. 앞의 5개 값 확인 (상위 5행)
df_books[['상품명', '발행일']].head()

발행일 컬럼 자료형: datetime64[us]


,상품명,발행일
0,소년이 온다,2014-05-19
1,모순,2013-04-01
2,결국 국민이 합니다,2025-04-15
3,혼모노,2025-03-28
4,급류,2022-12-22


### 발행일 전처리 실행 결과 요약

* **전처리 내용:** `.astype(str)`로 문자열 변환 후 `pd.to_datetime(..., format='%Y%m%d')`를 적용해 날짜 형식으로 변환 완료
* **자료형 확인:** `발행일` 컬럼의 자료형이 정수형(`int64`)에서 날짜형(`datetime64[us]`)으로 성공적으로 변경됨
* **데이터 확인:** 상위 5행 출력 결과, `20140519` $\rightarrow$ `2014-05-19`와 같이 `YYYY-MM-DD` 형태로 정확히 변환된 것을 확인

### 실습 8. 문자열 앞뒤 공백 정리하기

 

상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

 

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

 

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.

 

"소설"

"소설 "

 

이번에 정리할 컬럼

 

상품명

저자

출판사

분야

In [18]:
# 1. 공백을 제거할 문자열 컬럼 목록 지정
str_cols = ['상품명', '저자', '출판사', '분야']

# 2. 반복문을 사용해 각 컬럼의 앞뒤 공백 제거 (str.strip() 적용)
for col in str_cols:
    df_books[col] = df_books[col].astype(str).str.strip()

# 3. 전처리 후 해당 컬럼들의 상위 5행 데이터 확인
df_books[str_cols].head()

,상품명,저자,출판사,분야
0,소년이 온다,한강,창비,소설
1,모순,양귀자,쓰다,소설
2,결국 국민이 합니다,이재명,오마이북,정치/사회
3,혼모노,성해나,창비,소설
4,급류,정대건,민음사,소설


### 문자열 공백 제거 실행 결과 요약

* **대상 컬럼:** `상품명`, `저자`, `출판사`, `분야` (총 4개)
* **전처리 내용:** 반복문(`for`)을 활용해 지정한 컬럼들의 텍스트 데이터에 `.astype(str).str.strip()`을 일괄 적용
* **결과 확인:** 문자열 양끝의 불필요한 공백이 성공적으로 제거되었으며, `df_books[str_cols].head()`를 통해 상위 5행 데이터가 정상 정제되었음을 확인

### 실습 9. 중복 데이터 확인하기

 

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

 

먼저 무엇을 기준으로 중복인지 생각해야 합니다.

 

예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.

 

하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

 

따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

 

우선 상품을 구분하는 판매상품ID를 확인합니다.
 

<중요한 생각>

 

상품명이 같다

≠ 반드시 잘못된 중복 데이터다

 

중복 데이터는 먼저 의미를 확인한 뒤 처리합니다.

 

이번 단계에서는 중복 여부만 확인하고 원본 행을 임의로 삭제하지 않습니다.

 

In [19]:
# 1. 전체 행이 완전히 똑같은 중복 데이터 개수 확인
duplicated_all = df_books.duplicated().sum()
print("전체 행 중복 개수:", duplicated_all)

# 2. '판매상품ID'가 중복된 행 개수 확인
duplicated_id = df_books.duplicated(subset=['판매상품ID']).sum()
print("판매상품ID 중복 개수:", duplicated_id)

# 3. '상품명'이 중복된 행 개수 확인
duplicated_name = df_books.duplicated(subset=['상품명']).sum()
print("상품명 중복 개수:", duplicated_name)

전체 행 중복 개수: 0
판매상품ID 중복 개수: 0
상품명 중복 개수: 0


### 중복 데이터 확인 실행 결과 요약

* **전체 행 중복 개수:** 0건
* **판매상품ID 중복 개수:** 0건
* **상품명 중복 개수:** 0건
* **결론:** `duplicated()` 함수 검사 결과, 전체 행뿐만 아니라 고유 키 역할을 하는 `판매상품ID` 및 `상품명` 기준에서도 중복 데이터가 0건으로 확인되어 데이터의 유일성이 잘 유지되고 있습니다.

### 실습 10. 전처리 결과 검증하기

 

지금까지 여러 전처리를 수행했습니다.

 

마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.

 

확인할 내용

 

전체 행과 컬럼 개수

컬럼 이름

자료형

결측치 개수

판매가 예시

발행일 예시

중복 여부

앞의 5행

In [20]:
# 1. 데이터 크기 (행, 열 개수) 확인
print("1. 데이터 크기 (shape):", df_books.shape)

# 2. 컬럼 목록 확인
print("\n2. 컬럼 목록:", df_books.columns.tolist())

# 3. 컬럼별 자료형 확인
print("\n3. 컬럼별 자료형 (dtypes):")
print(df_books.dtypes)

# 4. 컬럼별 결측치 개수 확인
print("\n4. 컬럼별 결측치 개수:")
print(df_books.isna().sum())

# 5. '판매가' 상위 5개 값 확인
print("\n5. '판매가' 상위 5개 값:")
print(df_books['판매가'].head())

# 6. '발행일' 상위 5개 값 확인
print("\n6. '발행일' 상위 5개 값:")
print(df_books['발행일'].head())

# 7. '판매상품ID' 중복 개수 확인
print("\n7. '판매상품ID' 중복 개수:", df_books.duplicated(subset=['판매상품ID']).sum())

# 8. 최종 데이터프레임 상위 5행 출력
print("\n8. 최종 DataFrame 미리보기 (head):")
df_books.head()

1. 데이터 크기 (shape): (199, 8)

2. 컬럼 목록: ['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야']

3. 컬럼별 자료형 (dtypes):
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object

4. 컬럼별 결측치 개수:
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

5. '판매가' 상위 5개 값:
0    13500
1    11700
2    19800
3    16200
4    12600
Name: 판매가, dtype: int64

6. '발행일' 상위 5개 값:
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]

7. '판매상품ID' 중복 개수: 0

8. 최종 DataFrame 미리보기 (head):


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설


### 최종 검증 실행 결과 요약

* **데이터 규모 (shape):** 199행 × 8열
* **컬럼 목록:** `순위`, `판매상품ID`, `상품명`, `판매가`, `저자`, `출판사`, `발행일`, `분야`
* **자료형 (dtypes):** 
  * `순위`, `판매가`: 정수형(`int64`)
  * `발행일`: 날짜형(`datetime64[us]`)
  * `판매상품ID`, `상품명`, `저자`, `출판사`, `분야`: 문자형(`str`)
* **결측치 (isna):** 전 컬럼 0건 (누락 없음)
* **중복 데이터:** `판매상품ID` 중복 개수 0건
* **데이터 미리보기:** `판매가` 수치형 변환, `발행일` 날짜형 변환, 컬럼명 변경 및 문자열 공백 제거가 모두 정상 적용된 1위~5위 상위 5행 데이터 확인

 

### 실습 11. 전처리 데이터 저장하기

 

다음 Chapter에서는 이 데이터를 다시 사용합니다.

 

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

 

추천 파일명은 다음과 같습니다.

 

book_bestseller_clean.csv

 

한글 데이터가 Excel에서 깨지는 문제를 줄이기 위해 utf-8-sig 인코딩을 사용할 수 있습니다.

 

In [21]:
# 전처리가 완료된 df_books 데이터프레임을 CSV 파일로 저장
df_books.to_csv('book_bestseller_clean.csv', index=False, encoding='utf-8-sig')

### 교보문고 베스트셀러 데이터 기본 전처리 완료

#### 1. 무엇을 확인했는지
- **원본 데이터 크기:** (199, 8)
- **선택한 컬럼 수:** 8개
- **결측치 확인:** 저자 결측치 0개, 분야 결측치 0개
- **중복 데이터 확인:** 판매상품ID 중복 0개
- **자료형 확인:** 판매가 자료형(int64), 발행일 자료형(datetime64[us])

#### 2. 어떤 전처리를 했는지
- 분석에 필요한 컬럼만 선택하고, 컬럼 이름을 보기 편하게 변경했습니다.
- '판매가'의 쉼표(`,`)를 제거한 뒤 수치형 데이터로 변환했습니다.
- '발행일'을 날짜 형식(datetime)으로 변환했습니다.
- 결측치와 중복 데이터가 없는지 확인했습니다.

#### 3. 다음 분석을 위해 어떤 상태가 되었는지
- 데이터의 타입이 수치형과 날짜형으로 잘 정리되어 계산과 날짜별 분석이 가능한 상태가 되었습니다.
- 정제된 데이터를 `book_bestseller_clean.csv` 파일로 안전하게 저장했습니다.